### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by annanna**
Basic functionality for creating, reading, updating and deleting documents from Cloud Firestore and creating and reading files from Cloud Storage.
This repository is a work-in-progress, the package is integrated in all our existing projects after each other and functions are standardized. The snapshot-listeners for example come with the integration into the QTM-project, so this is going to be embedded in the next version.

### CODE REVIEW

File: .github/README.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# FirebaseManager Package
+
+This repository contains all utility functions that simplify transactions with Firebase products like Cloud Firestore, Cloud Storage or Authentication, wrapped in a Swift Package.
+It is made by **[SPACE SQUAD](https://www.spacesquad.de)**! We make great software with ♥️ in Berlin.
```
Comment (kaevinio) [inline .github/README.md:5]: I would prefer a separate line or paragraph for the SSQ mention, especially the logo should not be inline.

File: .github/README.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    let state: String?
+    let country: String?
+    let isCapital: Bool?
+    let population: Int64?
+
+    enum CodingKeys: String, CodingKey {
+        case name
+        case state
+        case country
+        case isCapital = "capital"
```
Comment (kaevinio) [inline .github/README.md:75]: Why do you define this separately as "capital"? Is this for example purposes?

File: .github/README.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+```Swift
+// Create new document
+let country = Country(....)
+FirestoreManager.createDocument(country, reference: FirestoreReference.country) { _ in }
+
+// Read all documents from a collection
+FirestoreManager.fetchCollection(FirestoreReference.country) { (result: Result<[Country], FirestoreError>) in
+    switch result {
+    case .success(let countries):
+        print(countries)
```
Comment (kaevinio) [inline .github/README.md:95]: We could add another comment to say "Here can use the fetched countries ..."

File: .github/README.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+// Update document
+country.name = "USA"
+FirestoreManager.updateDocument(country, reference: FirestoreReference.country, with: country.id) { _ in }
+
+// Delete document
+FirestoreManager.deleteDocument(reference: FirestoreReference.country, with: country.id)
+```
+
+### FirebaseStorageManager
+Target for all transactions with the Firebase Cloud Storage, using the Combine-framework.
```
Comment (kaevinio) [inline .github/README.md:110]: I love that it uses Combine! We don't use combine for Firestore yet, do we?

File: Sources/FirestoreManager/FirestoreCRUD/Fetch.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     
+     - Parameter reference: The collection name
+     - Parameter userFilter: Filter all objects by the current user's id
+     - Parameter filters: Dictionary of the filter key and the value
+     - Parameter orderBy: Key to order objects by
+     - Parameter descending: Whether orderBy key should descend
+     - Parameter limit: Limit number of fetched items
+     */
+    public static func fetchCollection<T>(_ reference: ReferenceProtocol,
+                                   userFilter: Bool = true,
```
Comment (kaevinio) [inline Sources/FirestoreManager/FirestoreCRUD/Fetch.swift]: The userFilter isn't called within the function. Either we need to call it or remove it. If we use it, there needs to be the property for the userID as well.

File: Sources/FirestoreManager/FirestoreCRUD/Update.swift [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        
+        do {
+            try reference.reference().document(id).setData(from: object, merge: false)
+            completion(.success(id))
+        } catch {
+            completion(.failure(.update(error: error)))
+        }
+    }
+    
+    public static func batchWrite<T>(_ data: [T], reference: ReferenceProtocol, parentId: String? = nil, completion: @escaping(FirestoreError?) -> Void) where T: Encodable {
```
Comment (kaevinio) [inline Sources/FirestoreManager/FirestoreCRUD/Update.swift:32]: Does this batch create or batch update? I think documentation for the function would be good, also to state what the parentId is, as this parameter isn't used somewhere else and might be confusing.

File: Sources/FirestoreManager/FirestoreError.swift [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+//  FirestoreError.swift
+//  
+//
+//  Created by Anna Münster on 03.06.22.
+//
+
+import Foundation
+
+public enum FirestoreError: LocalizedError {
+//    case notAuthorized
```
Comment (kaevinio) [inline Sources/FirestoreManager/FirestoreError.swift:11]: Why is this commented out?

File: Sources/FirestoreManager/FirestoreManager.swift [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+//
+//  Created by Anna Münster on 26.08.21.
+//
+
+import Firebase
+import FirebaseFirestoreSwift
+
+
+open class FirestoreManager {
+    static let database = Firestore.firestore()
```
Comment (kaevinio) [inline Sources/FirestoreManager/FirestoreManager.swift:13]: As there is is also Firebase Database (Realtime Database), I think renaming this to "firestore" would be better.

File: Sources/FirestoreManager/FirestoreReference.swift [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import Firebase
+
+public protocol ReferenceProtocol {
+    var rawValue: String { get }
+    var parent: ReferenceProtocol? { get }
+}
+
+extension ReferenceProtocol {
+    func reference(parentId: String? = nil) -> CollectionReference {
+        if let parent = parent, let parentId = parentId {
```
Comment (kaevinio) [inline Sources/FirestoreManager/FirestoreReference.swift:17]: I guess the parentId here is to prepare for nested fetching/creating, right? Maybe we can add a comment for now to make that clear.

File: Sources/StorageManager/FileType.swift [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+//
+//  Created by Anna Münster on 28.06.22.
+//
+
+import Foundation
+
+public enum FileType: String {
+    case video = "mp4"
+    case csv
+    case image = "jpg"
```
Comment (kaevinio) [inline Sources/StorageManager/FileType.swift:13]: What about png?

File: Sources/StorageManager/FirebaseStorageManager.swift [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+                if let error = error {
+                    promise(.failure(.upload(error: error)))
+                }
+                promise(.success(true))
+            }
+            
+            taskReference.observe(.progress) { snapshot in
+                guard let progress = snapshot.progress?.fractionCompleted
+                
+                else { return }
```
Comment (kaevinio) [inline Sources/StorageManager/FirebaseStorageManager.swift]: Is this weird line indentation from Github? If not, please correct it.

File: Sources/StorageManager/FirebaseStorageManager.swift [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+                let storageRef = Storage.storage().reference(withPath: "\(path)/\(fileName).\(fileType.rawValue)")
+                
+                storageRef.write(toFile: localFile) { url, error in
+                    if let error = error {
+                        promise(.failure(.download(error: error)))
+                        return
+                    }
+                    
+                    guard let url = url
+                    else {
```
Comment (kaevinio) [inline Sources/StorageManager/FirebaseStorageManager.swift]: Also the indentation here.

File: Package.swift [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    products: [
+        .library(
+            name: "FirebaseManagerPackage",
+            targets: [
+                "FirebaseStorageManager"
+            ])
+    ],
+    dependencies: [
+        .package(url: "https://github.com/firebase/firebase-ios-sdk.git",
+                 .branch("master"))
```
Comment (kaevinio) [inline Package.swift]: Do we want the master here? Maybe the safer option is up to next major level if there are any breaking changes?

File: Package.swift [hunk 14]
```diff
@@ -0,0 +1,39 @@
+// swift-tools-version:5.5
+// The swift-tools-version declares the minimum version of Swift required to build this package.
+
+import PackageDescription
+
+let package = Package(
+    name: "FirebaseManager",
+    platforms: [
+        .iOS(.v14),
+        .macOS(.v10_15)
```
Comment (kaevinio) [inline Package.swift]: We support iOS 14 but only the most recent version of macOS? Do you know whether there is a difference, so the package wouldn't work on a previous version? Or should we should just go for iOS 15, so we could use more recently supported functions as well?

**Review by kaevinio — CHANGES_REQUESTED** [review]

File: .github/README.md (hunk 4 above)
Comment (annanna) [reply .github/README.md:110]: No, not for now. But we could add this in future!
We could actually offer multiple options to let people decide what to use? (Completion, Combine, Async/Await)

File: Sources/FirestoreManager/FirestoreCRUD/Fetch.swift (hunk 5 above)
Comment (annanna) [reply Sources/FirestoreManager/FirestoreCRUD/Fetch.swift]: I removed it in here and will provide a new method signature probably in the next PR

File: Sources/StorageManager/FileType.swift (hunk 10 above)
Comment (annanna) [reply Sources/StorageManager/FileType.swift:13]: Is probably no problem either. I didn't test it in here and just wrote that this is to be extended. I thought it makes sense to add all file types on the fly when we need them. Or all at once

File: Sources/StorageManager/FirebaseStorageManager.swift (hunk 11 above)
Comment (annanna) [reply Sources/StorageManager/FirebaseStorageManager.swift]: Fixed

File: Sources/StorageManager/FirebaseStorageManager.swift (hunk 12 above)
Comment (annanna) [reply Sources/StorageManager/FirebaseStorageManager.swift]: Fixed

File: Package.swift (hunk 13 above)
Comment (annanna) [reply Package.swift]: Yes, changed it!

File: .github/README.md (hunk 1 above)
Comment (annanna) [reply .github/README.md:5]: Fixed

File: .github/README.md (hunk 2 above)
Comment (annanna) [reply .github/README.md:75]: This is sample is taken from the Firebase documentation and I found it valuable to point out here that different rawValues can be used here

File: .github/README.md (hunk 3 above)
Comment (annanna) [reply .github/README.md:95]: Good point, I added it

File: Package.swift (hunk 14 above)
Comment (annanna) [reply Package.swift]: Yes, changed macOS also to 14. Should be fine since the referenced packages support way older

**Review by kaevinio — APPROVED** [review]

**Pull request merged**